### ISSUE TITLE
ENH: Add a way for ufuncs with NEP 42 dtypes to do value-based promotion

### ISSUE THREAD

**Issue by ngoldbaum** [numpy/numpy#22809]
### Proposed new feature or change:

Currently NEP 42 dtypes do not have a way to determine the output dtype of an operation based on the values of the inputs.

I hit this issue trying to implement string multiplication for [`ASCIIDType`](https://github.com/numpy/numpy-user-dtypes/tree/main/asciidtype):

```python
import numpy as np
from asciidtype import ASCIIDType

dtype = ASCIIDType(5)

arr = np.array(["data"], dtype=dtype)

arr * 2
```

In this example, the output dtype should be `ASCIIDType(10)` (10 character wide ASCII), following what currently happens with `np.char` for string or byte dtypes:

```python
In [8]: arr = np.array(["hello", "world"], dtype="U5")

In [9]: np.char.multiply(arr, [2, 3])
Out[9]: array(['hellohello', 'worldworldworld'], dtype='<U15')

In [10]: np.char.multiply(arr, 2)
Out[10]: array(['hellohello', 'worldworld'], dtype='<U10')
```

Another example is quantity exponentiation:

```python
import numpy as np
from unitdtype import UnitDType

arr = np.array([1, 2, 3], dtype=UnitDType('meter'))
arr ** 2
```

The result should have units of `meter**2`, but we can't know that without inspecting the value of the second operand to the power ufunc. Note that unlike the string multiplication example, unit exponentiation only makes sense when exponentiating by a scalar.

NEP 42 defines the `resolve_descriptors` function to determine the output dtype for ufuncs, but that function only has access to the input dtypes, not the input values.

@seberg tells me that it was an explicit design decision to not allow looking at the input data in `resolve_descriptors` since it can enable confusing runtime behaviors. Since some use cases *need* this to work, a nice compromise might be to define an alternative to `NPY_METH_resolve_descriptors` that allows implementors to write a dtype resolution function that has pointers to the input arrays. That way dtype implementations will need to explicitly opt in to this behavior.

**Comment by seberg** [comment]
Yeah, it was an intentional design goal.  The `arr**2` was always the one thing that made me pause, but I thought that maybe `__pow__` is maybe special enough and can be special compared to `np.power` (which allows an array power, so things would be awkward for units anyway).

I still feel it is a good general design goal to have this as the normal way of doing things.  But I don't disagree that sometimes it would be pragmatic to have more power.
For strings, I do wonder whether it is worthwhile compared to having a `repeat()` function that is simply not quite a ufunc.  (Plus if we had arbitrary length stringth, this issue might evaporate.)

An other alternative might be to think of an "all or nothing" approach that is (unfortunately) NumPy specific (none of the current API is NumPy specific).  I.e. a way to take over completely from the ufunc machinery (similar to `__array_ufunc__`).

**Comment by mhvk** [comment]
The example of the `Quantity` is interesting -- astropy similarly needs to special-case this because indeed the result unit depends on the value. As is, we override `np.power` and have a special branch for it in `__array_ufunc__`, but I think in principle, just overriding `__pow__` would be reasonable (i.e., `np.power` would just not work for anything not dimensionless). However, how can the `dtype` decide how `__pow__` behaves? If it cannot, then one is forced to have a special `Quantity` class -- `ndarray` just couldn't be a real quantity even with a unit dtype. The easiest approach may be to allow for more power indeed...

**Comment by ngoldbaum** [comment]
Reductions are another place where it might be nice to be able to do value-based promotion. In particular, addition for strings and products for units have output dtype parameters that require looking at the input data. For reductions we could probably get away with `resolve_descriptors` only receiving the number of elements that will be reduced and not actually look at the data values.

**Comment by ngoldbaum** [comment]
I don't know if it's quite the same thing, since it's a cast, but I realized today that casting from a variable-width string dtype to `np.unicode_` can't work without something like this, since the size of the longest string can't be discovered inside the `resolve_descriptors` implementation for the cast. In practice this means that users must always specify a concrete size to cast to. It's not a problem for unicode to string cast.

**Comment by seberg** [comment]
Right, and currently I guess the only option you have is:
1. Raise an error in the resolution function.
2. Provide a small function:
   ```python
   def convert_to_np_unicode(arr):
      length = np.get_stringlength_ufunc(arr)
      return arr.astype(NumPyUnicode(length), casting="unsafe")
   ``

(which honestly I don't think is as bad as it might seem on first sight!  Users shouldn't convert from an arbitrary length to a fixed-length string often.)

That said, maybe we can find a way to make at least `arr.astype()` work.  If that was so, the ufunc machinery could make even string loops work in theory.  That wouldn't make `power` for units work unless you introduce a funny new dtype just for the purpose. 

The problem remains that avoiding value-based things is very useful for static analysis which is fundamental for accelerators or `dask.array` although maybe dask can workaround it.

**Comment by ngoldbaum** [comment]
Yeah, that small function would be helpful. I'll include that and mention it in the error message for now.

Another consequence I realized today is that the string to fixed-width unicode `resolve_descriptors` must pessimistically say the cast is unsafe, since there's no way to see whether the strings are actually short enough to fit into the output without truncation.
